Финал 11 спринта

ove2xrate opened the pull request:
Сдача на ревью

ove2xrate committed:
Подготовка проекта к ревью:
✅ переписан кодз-за проблем с синхронизацией с Гитом
✅ добавлено расширение UIColor для упрощенной работы с цветами.
✅ добавлен convertFromSnakeCase

helecta commented on image_feed/Auth/AuthViewController.swift:
```diff
@@ -1,31 +1,31 @@
-    weak var delegate: AuthViewControllerDelegate?
-    
-    // MARK: - Private Properties
     private let ShowWebViewSegueIdentifier = "ShowWebView"
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on image_feed/Auth/WebViewViewController.swift:
```diff
@@ -1,90 +1,81 @@
+    // MARK: - Private methods
     
     private func updateProgress() {
         progressView.progress = Float(webView.estimatedProgress)
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on image_feed/Helpers/URLRequest+Extension.swift:
```diff
@@ -0,0 +1,24 @@
+
+final class URLRequestBuilder {
+    
+    static let shared = URLRequestBuilder()
```
Предполагается, что URLRequestBuilder будет использоваться в одном экземпляре, т.е. через shared. Тогда стоит добавить приватный init. Так мы реализуем паттерн  [singleton](https://www.hackingwithswift.com/example-code/language/what-is-a-singleton) .

helecta commented on image_feed/OAuth2/OAuth2TokenStorage.swift:
```diff
@@ -1,15 +1,23 @@
 final class OAuth2TokenStorage {
-    let tokenKey = "BearerToken"
-    private let userDefaults = UserDefaults.standard
+    static let shared = OAuth2TokenStorage()
```
И тут стоит добавить приватный init. (паттерн синглтон)

helecta commented on image_feed/Profile/ProfileImageService.swift:
```diff
@@ -1,8 +1,66 @@
+final class ProfileImageService {
+    
+    // MARK: - Properties
+    static let shared = ProfileImageService()
```
И тут стоит добавить приватный init. (паттерн синглтон)

helecta commented on image_feed/Profile/ProfileService.swift:
```diff
@@ -1,8 +1,72 @@
+
+final class ProfileService {
+    
+    static let shared = ProfileService()
```
И тут стоит добавить приватный init. (паттерн синглтон)

helecta commented on image_feed/Profile/ProfileViewController.swift:
```diff
@@ -1,72 +1,152 @@
+    
+    private var descriptionLabel:UILabel = {
+        let label = UILabel()
+        label.font = UIFont.systemFont(ofSize: 13, weight: UIFont.Weight.regular)
```
Это поле может быть длинным, в этом случае у тебя текст будет обрезаться.
Лучше тут сделать многострочный вариант строки. Для этого нужно сделать так:
```
label.numberOfLines = 0
```
В этом случае еще нужно добавить trailingAnchor для labelText

helecta commented on image_feed/Profile/ProfileViewController.swift:
```diff
@@ -1,72 +1,152 @@
+            profileImage.heightAnchor.constraint(equalToConstant: 70),
+            
+            nameLabel.heightAnchor.constraint(equalToConstant: 18),
+            nameLabel.leadingAnchor.constraint(equalTo: profileImage.leadingAnchor),
```
Тут тоже лучше добавить trailingAnchor, чтобы если имя будет слишком длинным, то оно нормально обрезалось

helecta commented on image_feed/Profile/ProfileViewController.swift:
```diff
@@ -1,72 +1,152 @@
+            nameLabel.topAnchor.constraint(equalTo: profileImage.bottomAnchor, constant: 8),
+            
+            loginLabel.heightAnchor.constraint(equalToConstant: 18),
+            loginLabel.leadingAnchor.constraint(equalTo: profileImage.leadingAnchor),
```
Тут тоже лучше добавить trailingAnchor, чтобы если логин будет слишком длинным, то он нормально обрезался

helecta commented on image_feed/SplashView/SplashViewController.swift:
```diff
@@ -1,76 +1,138 @@
-            else { fatalError("Failed to prepare for \(showAuthenticationScreenSegueIdentifier)") }
-            viewController.delegate = self
+    private func authStatusChecker () {
+        guard !wasChecked else { return }
```
логика не совсем рабочая:
Первый запуск приложения, пользователь не авторизован:
1) происходит проверка authStatusChecker
2) wasChecked выставляется в true
3) открывается экран авторизации и пользователь авторизовываются в вебвью
4) вебвью закрывается
5) опять вызывается проверка authStatusChecker
6) дальше первой строчки не пройдет, т.к wasChecked уже true, благодаря этому не будет показан опять экран авторизации
7) т.к закрылся экран авторизации, вызывается fetchAuthToken
8) если в запросе происходит ошибка, показан алерт, но после закрытия алерта, пользователь продолжает висеть на сплеш скрине, хотя должен попасть повторно на экран авторизации.

helecta replied:
Очень хорошо, что ты пытаешься решить проблему повторного открытия экрана авторизации после того как пользователь авторизовался. Но вводя такой флаг wasChecked, по сути ты даешь использовать метод authStatusChecker только один раз, это не совсем гибкое решение.

Как вариант можно сделать так:
1. Добавить в UIBlockingProgressHUD свойство `isShowing`:
```
final class UIBlockingProgressHUD {
    
    static var isShowing: Bool = false
    
    private static var window: UIWindow? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first
        else {
            return nil
        }
        return window
    }

    static func show() {
        isShowing = true
        window?.isUserInteractionEnabled = false
        ProgressHUD.show()
    }

    static func dismiss() {
        isShowing = false
        window?.isUserInteractionEnabled = true
        ProgressHUD.dismiss()
    }
}
```
2. Во viewDidAppear добавить проверку(вместо проверки флага):
```
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        guard UIBlockingProgressHUD.isShowing == false else { return }
...
```

helecta commented on image_feed/Models/OAuthTokenResponseBody.swift:
```diff
@@ -0,0 +1,15 @@
+    let scope: String
+    let createdAt: Int
+    
+    enum CodingKeys: String, CodingKey {
```
Если ты используешь декодер с `keyDecodingStrategy = .convertFromSnakeCase`, то тогда тут надо удалить CodingKeys иначе будет ошибка парсинга.
https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  - настройки JSONDecoder

ove2xrate committed:
Внесены правки после ревью:
✅ пофикшено поведение алерта
✅ добавлены потерянные приватные init'ы
✅ изменен authStatusChecker:  проверка через флаг wasChecked, заменена на свойство isShowing в UIBlockingProgressHUD.
✅ удалены CodingKeys

ove2xrate committed:
Внесены правки после ревью:

✅ пофикшено поведение алерта
✅ добавлены потерянные приватные init'ы
✅ изменен authStatusChecker:  проверка через флаг wasChecked, заменена на свойство isShowing в UIBlockingProgressHUD.
✅ удалены CodingKeys
